doc/configuration.md
# Configuration

## General

Each instance of vec2checkd by default reads its configuration file from `/etc/vec2checkd/conf.d/<instance_name>.yaml`. But another location can be provided via the `--config` flag.

The overall structure of the configuration file is simple:

```yaml
---
# Parameters needed for the Prometheus API client.
prometheus: {}

# Parameters needed for the Icinga API client.
icinga: {}

# Set of configurations that map PromQL query results to Icinga (passive) check results.
mappings: {}
```

The content of each section is further explained below.

First of all some clarifications how PromQL query results are mapped to passive Icinga check results:

* The overall "exit status", that is the Icinga host/service state is determined by comparing the value of each vector in the PromQL query result set with warning and critical thresholds. The most significant individual status determines the overall status (e.g. when ONE value lies within a critical threshold, the overall exit status is => "DOWN" / "CRITICAL").
* A passive check result can update eiher an Icinga host object (when _only_ the host name was provided in the mapping) or a service object (when _both_ the host name and service name were provided).

### Prometheus

The `prometheus` section of the configuration has the following structure.

```yaml
prometheus:
  # The URL at which the server is reachable in order to execute queries against the HTTP API.
  # OPTIONAL, default: 'http://localhost:9090'
  # Example:
  host: 'https://prometheus.example.com:9090'

  # Specify proxy settings.
  # OPTIONAL.
  proxy: <proxy_section>
```

For details on proxy usage see the section on [proxy settings](configuration.md#proxy).

### Icinga

The Icinga API is more difficult to set up in that it _requires_ HTTPS and either HTTP Basic Auth or x509 authentication. Please refer to the [Icinga documentation](https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/) in order to set up the API and a user object with an adequate set of permissions.
Once the API is set up, configure the required parameters in the `icinga` section.

```yaml
icinga:
  # The URL at which the server is reachable in order to execute queries against the HTTP API. In this case HTTPS is required.
  # OPTIONAL, default: 'https://localhost:5665'
  # Example:
  host: 'https://satellite1.icinga.local'

  # Specify proxy settings.
  # OPTIONAL.
  proxy: <proxy_section>

  # If no trust relationship between the system and the self-signed Icinga root certificate has been established by some means, the location of the certificate must be provided here.
  # OPTIONAL.
  # Example:
  ca_cert: '/usr/local/share/ca-certificates/icinga.crt'
  
  authentication:
    # Valid authentication mechanims are HTTP Basic Auth and X.509.
    # REQUIRED.
    method: 'x509' | 'basic-auth'

    # Username and password for Basic authentication are only needed when method is set to 'basic-auth'.
    username: 'myuser'
    password: 'mypassword'

    # Paths to client certificate and key are only needed when method is set to 'x509'.
    # Make sure the files are owned/readable by the vec2check user.
    client_cert: '/var/lib/vec2checkd/ssl'
    client_key: '/var/lib/vec2checkd/ssl'
```

Note that the Icinga ApiUser username and password (Basic auth.) may also be read from the environment using the variables **V2C_ICINGA_USERNAME** and **V2C_ICINGA_PASSWORD** respectively. When the username and password are defined in both the environment and the configuration file, the values from the environment take precedence over the YAML parameters.

For details on proxy usage see the section on [proxy settings](configuration.md#proxy).

### Proxy

Proxy settings can be configured in the `prometheus` and `icinga` sections. This is the general structure:

```yaml
proxy:
  # Ignore any proxy settings set in this file or via environment variables.
  # OPTIONAL, default: false.
  ignore: true|false

  # Specify a proxy URL.
  # OPTIONAL.
  host: http://pro.xy:3128
```

By default both the Prometheus and Icinga clients read the common environment variables as well (that is HTTP_PROXY, HTTPS_PROXY, NO_PROXY and their lowercase pendants). When a proxy is specified in the configuration file the aforementioned environment variables are ignored.

### Mappings

A "mapping" defines a PromQL query to be executed and how to map the query result to a passive check result that is ultimately sent to the Icinga HTTP API.

```yaml
mappings:
  # Give each mapping a descriptive name as you would to a Prometheus recording or alerting rule.
  '<name>':
    # Probably self-explanatory. Specify a PromQL query to send to the Prometheus HTTP API.
    # REQUIRED.
    query: '<promql_query>'

    # The name of the Icinga host object to be updated.
    # REQUIRED.
    host: '<host_object'

    # The name of the Icinga service object to be updated.
    # Note: When this is omitted the passive check result will update the host object instead.
    # OPTIONAL.
    service: '<service_object>'

    # Check interval or how often á mapping is processed (in seconds). Must be in the range 10..=3600.
    # OPTIONAL, default 60.
    interval: <check_interval_in_seconds>

    # Use warning and critical thresholds to check every value in each time series in the PromQL result and determine the overall Icinga host/service state.
    # Each threshold must be a Nagios range.
    # OPTIONAL.
    thresholds:
      # OPTIONAL.
      warning: '<nagios_range>'

      # OPTIONAL.
      critical: '<nagios_range>'

    # Used to customize output when the default output does not suffice.
    # OPTIONAL.
    plugin_output: '<custom_output>'

    # Define if and how to send performance data as part of a passive check result.
    # OPTIONAL.
    performance_data:
      # Choose if performance data is sent.
      # OPTIONAL, default true.
      enabled: true|false

      # Customize the performance data label if desired.
      # OPTIONAL.
      label: '<custom_label>'

      # Customize the unit of measurement if desired.
      # OPTIONAL.
      uom: '<custom_unit_of_measurement>'
```

Some of these parameters are further explained in the following sections:

* On valid **Nagios ranges** see [Nagios development guidelines](https://nagios-plugins.org/doc/guidelines.html#THRESHOLDFORMAT) or [Icinga2 documentation](https://icinga.com/docs/icinga-2/latest/doc/05-service-monitoring/#threshold-ranges)
* On **plugin output** and customization see [this document](plugin_output.md)
* On **performance data** and customization see [this document](performance_data.md)



src/config.rs
use crate::error::*;
use crate::types::*;
use anyhow::{anyhow, bail};
use log::debug;
use nagios_range::NagiosRange;
use std::env;
use std::path::PathBuf;
use std::time::{Duration, Instant};
use url::Url;
use yaml_rust::yaml::{Hash, Yaml};

/// Parses a single mapping from YAML configuration.
/// This YAML is expected to have the following format:
///
/// ```yaml
/// '<name>':
///   query: '<promql_query>'
///   host: '<host_object>'
///   service: '<host_object>'           # optional
///   interval: <check_interval>
///   thresholds:                        # optional
///     warning: '<nagios_range>'        # optional
///     critical: '<nagios_range>'       # optional
///   plugin_output: '<custom_template>' # optional
/// ```
fn parse_mapping(mapping: (&Yaml, &Yaml)) -> Result<Mapping, anyhow::Error> {
    let name = mapping
        .0
        .as_str()
        .ok_or(ParseFieldError {
            field: "mappings.$name".to_string(),
            kind: "string",
        })?
        .to_string();

    let items = mapping.1.as_hash().ok_or(ParseFieldError {
        field: format!("mappings.{}", name),
        kind: "hash",
    })?;

    let query = items
        .get(&Yaml::from_str("query"))
        .ok_or(MissingFieldError {
            field: format!("mappings.{}.query", name),
        })?
        .as_str()
        .ok_or(ParseFieldError {
            field: format!("mappings.{}.query", name),
            kind: "string",
        })?
        .to_string();

    let host = items
        .get(&Yaml::from_str("host"))
        .ok_or(MissingFieldError {
            field: format!("mappings.{}.host", name),
        })?
        .as_str()
        .ok_or(ParseFieldError {
            field: format!("mappings.{}.host", name),
            kind: "string",
        })?
        .to_string();

    let service = match items.get(&Yaml::from_str("service")) {
        Some(s) => Some(
            s.as_str()
                .ok_or(ParseFieldError {
                    field: format!("mappings.{}.service", name),
                    kind: "string",
                })?
                .to_string(),
        ),
        None => None,
    };

    let plugin_output = match items.get(&Yaml::from_str("plugin_output")) {
        Some(p) => Some(
            p.as_str()
                .ok_or(ParseFieldError {
                    field: format!("mappings.{}.plugin_output", name),
                    kind: "string",
                })?
                .to_string(),
        ),
        None => None,
    };

    let thresholds = {
        match items.get(&Yaml::from_str("thresholds")) {
            Some(t) => {
                let t_hash = t.as_hash().ok_or(ParseFieldError {
                    field: format!("mappings.{}.thresholds", name),
                    kind: "hash",
                })?;

                if t_hash.is_empty() {
                    ThresholdPair::default()
                } else {
                    let warning = match t_hash.get(&Yaml::from_str("warning")) {
                        Some(w) => {
                            let w_raw = w.as_str().ok_or(ParseFieldError {
                                field: format!("mappings.{}.thresholds.warning", name),
                                kind: "string",
                            })?;
                            Some(NagiosRange::from(w_raw)?)
                        }
                        None => None,
                    };

                    let critical = match t_hash.get(&Yaml::from_str("critical")) {
                        Some(c) => {
                            let c_raw = c.as_str().ok_or(ParseFieldError {
                                field: format!("mappings.{}.thresholds.critical", name),
                                kind: "string",
                            })?;
                            Some(NagiosRange::from(c_raw)?)
                        }
                        None => None,
                    };

                    ThresholdPair { warning, critical }
                }
            }
            None => ThresholdPair::default(),
        }
    };

    let interval: Duration = match items.get(&Yaml::from_str("interval")) {
        Some(i) => {
            let num = i.as_i64().ok_or(ParseFieldError {
                field: format!("mappings.{}.interval", name),
                kind: "number",
            })?;

            let conv = u16::try_from(num).map_err(|_| ParseFieldError {
                field: format!("mappings.{}.interval", name),
                kind: "number",
            })? as u64;

            let valid_range = 10..=3600;

            if !valid_range.contains(&conv) {
                return Err(anyhow!(
                    "mappings.{}.interval must be in the range {:?}, got {}",
                    name,
                    valid_range,
                    conv
                ));
            }

            Duration::from_secs(conv)
        }
        None => Duration::from_secs(60),
    };

    let performance_data = {
        match items.get(&Yaml::from_str("performance_data")) {
            Some(t) => {
                let t_hash = t.as_hash().ok_or(ParseFieldError {
                    field: format!("mappings.{}.performance_data", name),
                    kind: "hash",
                })?;

                if t_hash.is_empty() {
                    PerformanceData::default()
                } else {
                    let enabled = match t_hash.get(&Yaml::from_str("enabled")) {
                        Some(val) => val.as_bool().ok_or(ParseFieldError {
                            field: format!("mappings.{}.performance_data.enabled", name),
                            kind: "boolean",
                        })?,
                        None => true,
                    };

                    let label = match t_hash.get(&Yaml::from_str("label")) {
                        Some(val) => {
                            let label = val
                                .as_str()
                                .ok_or(ParseFieldError {
                                    field: format!("mappings.{}.performance_data.label", name),
                                    kind: "string",
                                })?
                                .to_string();
                            Some(label)
                        }
                        None => None,
                    };

                    let uom = match t_hash.get(&Yaml::from_str("uom")) {
                        Some(val) => {
                            let label = val
                                .as_str()
                                .ok_or(ParseFieldError {
                                    field: format!("mappings.{}.performance_data.uom", name),
                                    kind: "string",
                                })?
                                .to_string();
                            Some(label)
                        }
                        None => None,
                    };

                    PerformanceData {
                        enabled,
                        label,
                        uom,
                    }
                }
            }
            None => PerformanceData::default(),
        }
    };

    Ok(Mapping {
        name,
        query,
        host,
        service,
        interval,
        plugin_output,
        thresholds,
        performance_data,
        last_apply: Instant::now(),
    })
}

/// Parses a multiple mappings from YAML configuration.
/// This YAML is expected to have the following format:
///
/// ```yaml
/// mappings:
///   '<first>': {} ...
///   '<second>': {} ...
///   '<third>': {} ...
///   ...
/// ```
pub(crate) fn parse_mappings(config: Hash) -> Result<Vec<Mapping>, anyhow::Error> {
    let mut mappings: Vec<Mapping> = vec![];

    match config.get(&Yaml::from_str("mappings")) {
        Some(m_raw) => {
            let mapping_hash = m_raw.as_hash().ok_or(ParseFieldError {
                field: String::from("mappings"),
                kind: "hash",
            })?;

            for raw_mapping in mapping_hash {
                let mapping = parse_mapping(raw_mapping)?;
                mappings.push(mapping);
            }

            Ok(mappings)
        }
        None => Ok(vec![]),
    }
}

fn parse_proxy_section(config: &Hash, scheme: &str) -> Result<ProxyConfig, anyhow::Error> {
    let ignore = config
        .get(&Yaml::from_str("ignore"))
        .unwrap_or(&Yaml::Boolean(false))
        .as_bool()
        .ok_or(ParseFieldError {
            field: String::from("proxy.ignore"),
            kind: "boolean",
        })?;

    let host = match config.get(&Yaml::from_str("host")) {
        Some(val) => {
            let raw = val.as_str().ok_or(ParseFieldError {
                field: String::from("proxy.host"),
                kind: "string",
            })?;
            match scheme {
                "http" => Some(reqwest::Proxy::http(raw)?),
                "https" => Some(reqwest::Proxy::https(raw)?),
                _ => unreachable!(),
            }
        }
        None => None,
    };

    Ok(ProxyConfig { ignore, host })
}

pub(crate) fn parse_prom_section(config: &Hash) -> Result<PromConfig, anyhow::Error> {
    let default_host = Url::parse("http://localhost:9090")?;

    match config.get(&Yaml::from_str("prometheus")) {
        Some(section) => {
            let prometheus = section.as_hash().ok_or(ParseFieldError {
                field: String::from("prometheus"),
                kind: "hash",
            })?;

            let host = match prometheus.get(&Yaml::from_str("host")) {
                Some(h) => {
                    let raw = h.as_str().ok_or(ParseFieldError {
                        field: String::from("prometheus.host"),
                        kind: "string",
                    })?;
                    let url = Url::parse(raw)?;
                    match url.scheme() {
                        "http" | "https" => url,
                        _ => bail!("the scheme in URL '{}' from field 'prometheus.host' must be either 'http' or 'https'", url)
                    }
                }
                None => default_host,
            };

            let proxy = match prometheus.get(&Yaml::from_str("proxy")) {
                Some(p) => p
                    .as_hash()
                    .ok_or(ParseFieldError {
                        field: String::from("prometheus.proxy"),
                        kind: "hash",
                    })
                    .map_err(anyhow::Error::msg)
                    .and_then(|s| parse_proxy_section(s, host.scheme()))?,
                None => ProxyConfig::default(),
            };

            Ok(PromConfig { host, proxy })
        }
        None => Ok(PromConfig {
            host: default_host,
            proxy: ProxyConfig::default(),
        }),
    }
}

pub(crate) fn parse_icinga_section(config: &Hash) -> Result<IcingaConfig, anyhow::Error> {
    let section = {
        let conf_attr = "icinga";
        config
            .get(&Yaml::from_str(conf_attr))
            .ok_or(MissingFieldError {
                field: String::from(conf_attr),
            })?
            .as_hash()
            .ok_or(ParseFieldError {
                field: String::from(conf_attr),
                kind: "hash",
            })?
    };

    let host = match section.get(&Yaml::from_str("host")) {
        Some(h) => {
            let raw = h.as_str().ok_or(ParseFieldError {
                field: String::from("icinga.host"),
                kind: "string",
            })?;
            let url = Url::parse(raw)?;
            match url.scheme() {
                "http" | "https" => url,
                _ => bail!("the scheme in URL '{}' from field 'prometheus.host' must be either 'http' or 'https'", url)
            }
        }
        None => Url::parse("https://localhost:5665")?,
    };

    let proxy = match section.get(&Yaml::from_str("proxy")) {
        Some(p) => p
            .as_hash()
            .ok_or(ParseFieldError {
                field: String::from("icinga.proxy"),
                kind: "hash",
            })
            .map_err(anyhow::Error::msg)
            .and_then(|s| parse_proxy_section(s, host.scheme()))?,
        None => ProxyConfig::default(),
    };

    let ca_cert = match section.get(&Yaml::from_str("ca_cert")) {
        Some(cert) => Some(
            cert.as_str()
                .ok_or(ParseFieldError {
                    field: String::from("icinga.ca_cert"),
                    kind: "string",
                })
                .map(PathBuf::from)?,
        ),
        None => None,
    };

    let auth_hash = {
        let conf_attr = "icinga.authentication";
        section
            .get(&Yaml::from_str("authentication"))
            .ok_or(MissingFieldError {
                field: String::from(conf_attr),
            })?
            .as_hash()
            .ok_or(ParseFieldError {
                field: String::from(conf_attr),
                kind: "hash",
            })?
    };

    let auth_method = {
        let conf_attr = "icinga.authentication.method";
        auth_hash
            .get(&Yaml::from_str("method"))
            .ok_or(MissingFieldError {
                field: String::from(conf_attr),
            })?
            .as_str()
            .ok_or(ParseFieldError {
                field: String::from(conf_attr),
                kind: "string",
            })?
    };

    let authentication = match auth_method {
        "basic-auth" => {
            let username = {
                let env_var = "V2C_ICINGA_USERNAME";
                let conf_attr = "icinga.authentication.username";

                match env::var(env_var) {
                    Ok(val) => val,
                    Err(err) => {
                        debug!("failed to read Icinga ApiUser username from environment: {err}; try to read from configuration file instead");

                        auth_hash
                            .get(&Yaml::from_str("username"))
                            .ok_or(anyhow!("failed to read mandatory Icinga ApiUser username from either the environment ('{env_var}') or configuration file ('{conf_attr}')")
                            )?
                            .as_str()
                            .ok_or(ParseFieldError {
                                field: conf_attr.to_string(),
                                kind: "string",
                            })?
                            .to_string()
                    }
                }
            };

            let password = {
                let env_var = "V2C_ICINGA_PASSWORD";
                let conf_attr = "icinga.authentication.password";

                match env::var(env_var) {
                    Ok(val) => val,
                    Err(err) => {
                        debug!("failed to read Icinga ApiUser password from environment: {err}; try to read from configuration file instead");

                        auth_hash
                            .get(&Yaml::from_str("password"))
                            .ok_or(anyhow!("failed to read mandatory Icinga ApiUser username from either the environment ('{env_var}') or configuration file ('{conf_attr}')")
                            )?
                            .as_str()
                            .ok_or(ParseFieldError {
                                field: conf_attr.to_string(),
                                kind: "string",
                            })?
                            .to_string()
                    }
                }
            };

            IcingaAuth::Basic(IcingaBasicAuth { username, password })
        }
        "x509" => {
            let client_cert = {
                let conf_attr = "icinga.authentication.client_cert";
                auth_hash
                    .get(&Yaml::from_str("client_cert"))
                    .ok_or(MissingFieldError {
                        field: String::from(conf_attr),
                    })?
                    .as_str()
                    .ok_or(ParseFieldError {
                        field: String::from(conf_attr),
                        kind: "string",
                    })
                    .map(PathBuf::from)?
            };

            let client_key = {
                let conf_attr = "icinga.authentication.client_key";
                auth_hash
                    .get(&Yaml::from_str("client_key"))
                    .ok_or(MissingFieldError {
                        field: String::from(conf_attr),
                    })?
                    .as_str()
                    .ok_or(ParseFieldError {
                        field: String::from(conf_attr),
                        kind: "string",
                    })
                    .map(PathBuf::from)?
            };

            IcingaAuth::X509(IcingaX509Auth {
                client_cert,
                client_key,
            })
        }
        _ => {
            bail!(
                    "invalid value in 'icinga.authentication.method', must be either 'basic-auth' or 'x509'"
                )
        }
    };

    Ok(IcingaConfig {
        host,
        ca_cert,
        authentication,
        proxy,
    })
}

pub(crate) fn parse_yaml(source: &str) -> Result<Hash, anyhow::Error> {
    yaml_rust::yaml::YamlLoader::load_from_str(source)?[0]
        .clone()
        .into_hash()
        .ok_or(anyhow!("failed to parse configuration as hash"))
}
